### PR TITLE
Handle ccall callee in inlining pass

### LINKED ISSUE #17403: Bad codegen for ccall callee
Noticed while checking https://github.com/JuliaLang/julia/issues/17395

`f1` and `f2` should basically be equivalent but apparently not. Type inference got it right but somehow now the codegen. (I guess this would be solved trivially with a linear IR?)

```jl
julia> const slot = Ref{Ptr{Void}}()
Base.RefValue{Ptr{Void}}(Ptr{Void} @0x00000000018136d0)                           
                                                                                  
julia> function f1()
           ccall(slot[], Void, ())
       end
f1 (generic function with 1 method)                                               
                                                                                  
julia> function f2()
           ptr = slot[]
           ccall(ptr, Void, ())
       end
f2 (generic function with 1 method)                                               
                                                                                  
julia> @code_llvm f1()
                                                                                  
define void @julia_f1_67762() #0 !dbg !6 {                                        
top:                                                                              
  %thread_ptr = call i8* asm "movq %fs:0, $0", "=r"() #4                          
  %ptls_i8 = getelementptr i8, i8* %thread_ptr, i64 -2672                         
  %ptls = bitcast i8* %ptls_i8 to %jl_value_t***                                  
  %0 = alloca [5 x %jl_value_t*], align 8                                         
  %.sub = getelementptr inbounds [5 x %jl_value_t*], [5 x %jl_value_t*]* %0, i64 0, i64 0                                                                           
  %1 = getelementptr [5 x %jl_value_t*], [5 x %jl_value_t*]* %0, i64 0, i64 3     
  %2 = getelementptr [5 x %jl_value_t*], [5 x %jl_value_t*]* %0, i64 0, i64 2     
  %3 = bitcast %jl_value_t** %1 to i8*
  call void @llvm.memset.p0i8.i32(i8* %3, i8 0, i32 16, i32 8, i1 false)
  %4 = bitcast [5 x %jl_value_t*]* %0 to i64*
  store i64 6, i64* %4, align 8
  %5 = getelementptr [5 x %jl_value_t*], [5 x %jl_value_t*]* %0, i64 0, i64 1
  %6 = bitcast i8* %ptls_i8 to i64*
  %7 = load i64, i64* %6, align 8
  %8 = bitcast %jl_value_t** %5 to i64*
  store i64 %7, i64* %8, align 8
  store %jl_value_t** %.sub, %jl_value_t*** %ptls, align 8
  store %jl_value_t* null, %jl_value_t** %2, align 8
  %9 = getelementptr [5 x %jl_value_t*], [5 x %jl_value_t*]* %0, i64 0, i64 4
  store %jl_value_t* inttoptr (i64 140728624867960 to %jl_value_t*), %jl_value_t** %1, align 8
  store %jl_value_t* inttoptr (i64 140728704309488 to %jl_value_t*), %jl_value_t** %9, align 8
  %10 = call %jl_value_t* @jl_apply_generic(%jl_value_t** %1, i32 2)
  store %jl_value_t* %10, %jl_value_t** %2, align 8
  %11 = bitcast %jl_value_t* %10 to i64*
  %12 = load i64, i64* %11, align 16
  %13 = icmp eq i64 %12, 0
  br i1 %13, label %fail, label %pass

fail:                                             ; preds = %top
  %14 = load %jl_value_t*, %jl_value_t** @jl_undefref_exception, align 8
  call void @jl_throw(%jl_value_t* %14)
  unreachable

pass:                                             ; preds = %top
  %15 = inttoptr i64 %12 to void ()*
  call void %15()
  %16 = load i64, i64* %8, align 8
  store i64 %16, i64* %6, align 8
  ret void
}

julia> @code_llvm f2()

define void @julia_f2_67766() #0 !dbg !6 {
top:
  %thread_ptr = call i8* asm "movq %fs:0, $0", "=r"() #3
  %0 = load i64, i64* inttoptr (i64 140728704309488 to i64*), align 16
  %1 = icmp eq i64 %0, 0
  br i1 %1, label %fail, label %pass

fail:                                             ; preds = %top
  %2 = load %jl_value_t*, %jl_value_t** @jl_undefref_exception, align 8
  call void @jl_throw(%jl_value_t* %2)
  unreachable

pass:                                             ; preds = %top
  %3 = inttoptr i64 %0 to void ()*
  call void %3()
  ret void
}
```

### PR DESCRIPTION
**Pull request by yuyichao**
Fix #17403

### CODE REVIEW

**Comment by yuyichao** [comment]
I saw the comment above but it seems that the inlining pass does not really mess with the const global case? Maybe I can add a check that the first argument is not a `Const` to be more robust?

**Comment by vtjnash** [comment]
codegen can handle linear-IR and nested IR, but not general inlining. right now, we aren't likely to inline something too complicated to handle. But it might be better to do this transform in the frontend instead anyways

File: base/inference.jl [hunk 1]
```diff
@@ -2965,7 +2965,8 @@ function inlining_pass(e::Expr, sv, linfo)
     # by the interpreter and inlining might put in something it can't handle,
     # like another ccall (or try to move the variables out into the function)
     if is_known_call(e, Core.Intrinsics.ccall, sv)
-        i0 = 5
+        # 4 is rewrite to 2 below to handle the callee.
```
Comment (tkelman) [inline base/inference.jl]: do you mean rewritten? ([ci skip] strictly typo-fix commits if you've verified it passes `make check-whitespace` locally)

**Comment by yuyichao** [comment]
> in the frontend

In the frontend without fully linear ir?

**Comment by vtjnash** [comment]
yes, to have the front-end to always move non-literals out of this statement position (the logic for this, iiuc, is already there, just guarded by full-literal-mode right now)

**Comment by yuyichao** [comment]
The logic is in a different pass and I don't think there's a easy way to turn that on only for `ccall` callee without duplicating the logic. I also assume that we don't want to turn very-linear-mode on right now....

**Pull request merged**